server/driver/wivrn_uinput.cpp

```diff
@@ -0,0 +1,232 @@
+#include "wivrn_uinput.h"
+
+#include <array>
+#include <cstdio>
+#include <cstring>
+
+#include <fcntl.h>
+#include <memory>
+#include <sys/ioctl.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include <linux/input-event-codes.h>
+#include <linux/uinput.h>
+
+namespace
+{
+int open_uinput_node()
+{
+	constexpr std::array<const char *, 2> paths = {"/dev/uinput", "/dev/input/uinput"};
+	for (const char * p: paths)
+	{
+		int fd = ::open(p, O_WRONLY | O_NONBLOCK);
+		if (fd >= 0)
+			return fd;
+	}
+	return -1;
+}
+
+void fill_ids(uinput_user_dev & uidev, uint16_t product)
+{
+	std::memset(&uidev, 0, sizeof(uidev));
+	uidev.id.bustype = 0x03;
+	uidev.id.vendor = 0x4711;
+	uidev.id.product = product;
+	uidev.id.version = 5;
+}
+
+bool emit_ev(int fd, uint16_t type, uint16_t code, int32_t value) noexcept
+{
+	struct input_event ev;
+	std::memset(&ev, 0, sizeof(ev));
+	gettimeofday(&ev.time, nullptr);
+	ev.type = type;
+	ev.code = code;
+	ev.value = value;
+	if (::write(fd, &ev, sizeof(ev)) != (ssize_t)sizeof(ev))
+	{
+		::close(fd);
+		return false;
+	}
+	return true;
+}
+
+bool syn(int fd) noexcept
+{
+	return emit_ev(fd, EV_SYN, SYN_REPORT, 0);
+}
+} // namespace
+
+std::unique_ptr<wivrn_uinput> wivrn_uinput::create()
+{
+	std::unique_ptr<wivrn_uinput> instance(new (std::nothrow) wivrn_uinput());
+	if (!instance)
+		return nullptr;
+	if (!instance->init_keyboard())
+		return nullptr;
+	if (!instance->init_mouse())
+		return nullptr;
+	return instance;
+}
+
+wivrn_uinput::wivrn_uinput() = default;
+
```

Comment:
Do the initialization in the constructor, don't be afraid to throw in case of error, catch in wivrn_session

Reply:
It is possible to forward errors to dbus/gui through https://github.com/WiVRn/WiVRn/blob/master/server/wivrn_ipc.h, see https://github.com/WiVRn/WiVRn/blob/master/server/driver/wivrn_comp_target.cpp#L329 for an example usage